Alignment/include/ActsAlignment/Kernel/detail/AlignmentEngine.hpp

```diff
@@ -0,0 +1,302 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Alignment.hpp"
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+#include "Acts/EventData/TrackParameters.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "ActsAlignment/Kernel/AlignmentMask.hpp"
+
+#include <unordered_map>
+
+namespace ActsAlignment {
+namespace detail {
+using namespace Acts;
+///
+///@brief struct to store info needed for track-based alignment
+///
+struct TrackAlignmentState {
+  // The dimension of measurements
+  size_t measurementDim = 0;
+
+  // The dimension of track parameters
+  size_t trackParametersDim = 0;
+
+  // The contributed alignment degree of freedom
+  size_t alignmentDof = 0;
+
+  // The measurements covariance
+  ActsDynamicMatrix measurementCovariance;
+
+  // The track parameters covariance
+  ActsDynamicMatrix trackParametersCovariance;
+
+  // The projection matrix
+  ActsDynamicMatrix projectionMatrix;
+
+  // The residual
+  ActsDynamicVector residual;
+
+  // The covariance of residual
+  ActsDynamicMatrix residualCovariance;
+
+  // The chi2
+  double chi2 = 0;
+
+  // The derivative of residual w.r.t. alignment parameters
+  ActsDynamicMatrix alignmentToResidualDerivative;
+
+  // The derivative of chi2 w.r.t. alignment parameters
+  ActsDynamicVector alignmentToChi2Derivative;
+
+  // The second derivative of chi2 w.r.t. alignment parameters
+  ActsDynamicMatrix alignmentToChi2SecondDerivative;
+
+  // The alignable surfaces on the track and their indices in both the global
+  // alignable surfaces pool and those relevant with this track
+  std::unordered_map<const Surface*, std::pair<size_t, size_t>> alignedSurfaces;
+};
+
```

Comment:
A question to this: since this is "per track" - what's the Dynamic size range of the Matrix & Vectors?

Reply:
For a single track, the size of the Matrix & Vectors is related to how many alignable detector objects this track is passing through, so it's expected to be at the order of O(1)*`eAlignmentSize`. For example, a track might be found to have measurements on 10 detector layers and 2 of the layers are requested to be aligned, and all the 6 alignment dof are float, then the size of the Matrix & Vectors for this track is 2*6=12; If we only align the center position of the two layers, then the size is 2*3=6.

Reply:
Ah, I see - I forgot the configuration aspect in this.